dali/kernels/imgproc/convolution/convolution_cpu.h

```diff
@@ -0,0 +1,362 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_CONVOLUTION_CONVOLUTION_CPU_H_
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_CONVOLUTION_CPU_H_
+
+#include "dali/core/boundary.h"
+#include "dali/core/convert.h"
+#include "dali/core/format.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/kernel.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+
+namespace dali {
+namespace kernels {
+/**
+ * @brief Cyclic buffer used for storing input window for convolution.
+ *
+ * Keeps `num_lanes` of each window element, intended when there is stride between
+ * window elements, but the same convolution can be applied to neighbouring innermost elements.
+ *
+ * Wraps a pointer to a buffer of type T, with size equal to length * num_lanes.
+ */
+template <typename T, int max_lanes = 16>
+class CyclicWindowWrapper {
+ public:
+  CyclicWindowWrapper(T* __restrict__ ptr, int length, int num_lanes = 1)
+      : data_(ptr), start_(0), end_(0), elements_(0), length_(length), num_lanes_(num_lanes) {
+    DALI_ENFORCE(num_lanes <= max_lanes, "Static lanes limit exceeded.");
+  }
+
+  /**
+   * @brief Drop one element consistine of `NumLanes()` lanes from the buffer, moving the start
+   *        element.
+   */
+  void PopElement() {
+    assert(elements_ > 0);
+    elements_--;
+    start_++;
+    WrapPosition(start_);
+  }
+
+  /**
+   * @brief Add a window element, that has `NumLanes()` consecutive lanes and starting at `input`,
+   * to the buffer.
+   */
+  void PushElement(const T* __restrict__ input) {
+    assert(elements_ < length_);
+    memcpy(data_ + end_ * NumLanes(), input, sizeof(T) * NumLanes());
+    elements_++;
+    end_++;
+    WrapPosition(end_);
+  }
+
+  /**
+   * @brief Add a window element to the buffer, that has `NumLanes()` lanes
+   */
+  void PushElement(span<const T> input) {
+    assert(elements_ < length_);
+    for (int c = 0; c < NumLanes(); c++) {
```

Comment:
do you need those two `PushElement` definitions or can we invoke one from the other ?

Reply:
I used it for the border, probably can remove it for now, it should be reworked to something else.